### PR TITLE
feat(aws): add aws lambdas and infra

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Setup the indexing service to run in AWS by building Lamdas for all the core functionality. All that's left is the terraform to define and ship the services

# Implementation

- A few more modifications to construct.Construct to make it more flexible
- Make notifier's HeadState work of abstract store.Store rather than datastore.Batching
- Make Notifier's main update logic callable seperately from the go routine timing loop
- Implement:
   - S3 implementation of store.Store
   - Dynamo implementation of store.ProviderContextTable
   - SQS implementation of providercacher.CachingQueue
   - SNS implementation of notifier.NotifyRemoteSyncFunc
   - Function to pull relevant AWS config from the environment
   - Construct function tailored to building the indexing service for AWS
   - Lambdas for:
      - All HTTP handlers
      - Provider caching from SQS event
      - Head state update checking on an EventBridge scheduler
      - RemoteSyncing from an SNS event

### CODE REVIEW

File: pkg/aws/s3store.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	}
+	return outPut.Body, nil
+}
+
+// Put implements store.Store.
+func (s *S3Store) Put(ctx context.Context, key string, data io.Reader) error {
+	_, err := s.s3Client.PutObject(ctx, &s3.PutObjectInput{
+		Bucket: aws.String(s.bucket),
+		Key:    aws.String(s.keyPrefix + key),
+		Body:   data,
```
Comment (alanshaw) [inline pkg/aws/s3store.go:45]: Is this ok? I thought you need to know length if streaming...

File: pkg/aws/snsremotesyncnotifer.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	Head string `json:"Head,omitempty"`
+	Prev string `json:"Prev,omitempty"`
+}
+
+type SNSRemoteSyncNotifier struct {
+	topicArn  string
+	snsClient *sns.Client
+}
+
+func NewSQSRemoteSyncNotifier(config aws.Config, topicArn string) *SNSRemoteSyncNotifier {
```
Comment (alanshaw) [inline pkg/aws/snsremotesyncnotifer.go:25]:
```suggestion
func NewSNSRemoteSyncNotifier(config aws.Config, topicArn string) *SNSRemoteSyncNotifier {
```

File: pkg/construct/construct.go [hunk 3]
```diff
@@ -29,7 +29,7 @@ import (
 
 var log = logging.Logger("service")
 var providerIndexNamespace = datastore.NewKey("providerindex/")
-var providerIndexPublisherNamespace = providerIndexNamespace.ChildString("publisher/")
+var providerIndexPublisherNamespace = providerIndexNamespace.Child(datastore.NewKey("publisher/"))
```
Comment (alanshaw) [inline pkg/construct/construct.go:32]: Why did change?

File: pkg/construct/construct.go [hunk 4]
```diff
@@ -85,6 +87,24 @@ func SkipNotification() Option {
 	}
 }
 
+// WithStartIPNIServer determines when IPNI adds will be served directly over HTTP
```
Comment (alanshaw) [inline pkg/construct/construct.go:90]:
```suggestion
// WithStartIPNIServer determines when IPNI ads will be served directly over HTTP
```

**Review by alanshaw — APPROVED** [review]
🚀 incredible work

File: pkg/construct/construct.go (hunk 3 above)
Comment (hannahhoward) [reply pkg/construct/construct.go:32]: childstring is a bit error prone on slashes

**Pull request merged**